api/client/commands.go

```diff
@@ -139,10 +143,42 @@ func (cli *DockerCli) CmdBuild(args ...string) error {
 		if _, err := os.Stat(root); err != nil {
 			return err
 		}
-		filename := path.Join(root, "Dockerfile")
+
+		absRoot, err := filepath.Abs(root)
+		if err != nil {
+			return err
+		}
+
+		var filename string       // path to Dockerfile
+		var origDockerfile string // used for error msg
+
+		if *dockerfileName == "" {
+			// No -f/--file was specified so use the default
+			origDockerfile = "Dockerfile"
+			*dockerfileName = origDockerfile
+			filename = path.Join(absRoot, *dockerfileName)
+		} else {
+			origDockerfile = *dockerfileName
+			if filename, err = filepath.Abs(*dockerfileName); err != nil {
+				return err
+			}
+		}
+
+		if !strings.HasSuffix(absRoot, "/") {
```

Comment:
Apparently filepath.Rel() will try really really hard to find a path that works - even if it has to generate one with ../../../.. to do it - notice the description even says:
On success, the returned path will always be relative to basepath, even if basepath and targpath *share no elements*

which doesn't make it nearly as useful.  Reverting parts of that change but open to other ideas.